planner/core/plan_cost_detail.go

```diff
@@ -134,3 +176,175 @@ func setPhysicalIndexReaderCostDetail(p *PhysicalIndexReader, opt *physicalOptim
 		RowCountLbl, RowSizeLbl, NetworkFactorLbl, NetSeekCostLbl, ScanConcurrencyLbl))
 	opt.appendPlanCostDetail(detail)
 }
+
+func setPhysicalHashJoinCostDetail(p *PhysicalHashJoin, opt *physicalOptimizeOp, spill bool,
+	buildCnt, probeCnt, cpuFactor, rowSize, numPairs,
+	cpuCost, probeCPUCost, memCost, diskCost, probeDiskCost,
+	diskFactor, memoryFactor, concurrencyFactor float64,
+	memQuota int64) {
+	if opt == nil {
+		return
+	}
+	detail := tracing.NewPhysicalPlanCostDetail(p.ID(), p.TP())
+	diskCostDetail := &HashJoinDiskCostDetail{
+		Spill:           spill,
+		UseOuterToBuild: p.UseOuterToBuild,
+		BuildRowCount:   buildCnt,
+		DiskFactor:      diskFactor,
+		RowSize:         rowSize,
+		ProbeDiskCost: &HashJoinProbeDiskCostDetail{
+			SelectionFactor: SelectionFactor,
+			NumPairs:        numPairs,
+			HasConditions:   len(p.LeftConditions)+len(p.RightConditions) > 0,
+			Cost:            probeDiskCost,
+		},
+		Cost: diskCost,
+	}
+	memoryCostDetail := &HashJoinMemoryCostDetail{
+		Spill:         spill,
+		MemQuota:      memQuota,
+		RowSize:       rowSize,
+		BuildRowCount: buildCnt,
+		MemoryFactor:  memoryFactor,
+		Cost:          memCost,
+	}
+	cpuCostDetail := &HashJoinCPUCostDetail{
+		BuildRowCount:     buildCnt,
+		CPUFactor:         cpuFactor,
+		ConcurrencyFactor: concurrencyFactor,
+		ProbeCost: &HashJoinProbeCostDetail{
+			NumPairs:        numPairs,
+			HasConditions:   len(p.LeftConditions)+len(p.RightConditions) > 0,
+			SelectionFactor: SelectionFactor,
+			ProbeRowCount:   probeCnt,
+			Cost:            probeCPUCost,
+		},
+		HashJoinConcurrency: p.Concurrency,
+		Spill:               spill,
+		Cost:                cpuCost,
+	}
+
+	// record cpu cost detail
+	detail.AddParam(CPUCostDetailLbl, cpuCostDetail).
+		AddParam(CPUCostDescLbl, cpuCostDetail.desc()).
+		AddParam(ProbeCostDescLbl, cpuCostDetail.probeCostDesc())
+	// record memory cost detail
+	detail.AddParam(MemCostDetailLbl, memoryCostDetail).
+		AddParam(MemCostDescLbl, memoryCostDetail.desc())
+	// record disk cost detail
+	detail.AddParam(DiskCostDetailLbl, diskCostDetail).
+		AddParam(DiskCostDescLbl, diskCostDetail.desc()).
+		AddParam(ProbeDiskCostDescLbl, diskCostDetail.probeDesc())
+
+	detail.SetDesc(fmt.Sprintf("%s+%s+%s+all children cost", CPUCostDetailLbl, MemCostDetailLbl, DiskCostDetailLbl))
```

Comment:
I notice that in previous PRs, each label in the `Desc` corresponds to exactly one value in the `Param` with the same label. But this PR doesn't follow this.
For example, we didn't explain the relationship between `CPUCostDetailLbl`, `CPUCostDescLbl`, and `ProbeCostDescLbl`.
Would we improve this sometime?

Reply:
Currently no, we may ignore this in interval server and revise it in dashboard.